### PR TITLE
Undertaker - First Pass

### PR DESCRIPTION
**Pull request by python-42**
PR for #9

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -57,4 +57,9 @@ public class Controller {
         public static final int LEFT_JOYSTICK_BUTTON = 9;
         public static final int RIGHT_JOYSTICK_BUTTON = 10;
     }
+
+    public class Undertaker {
+        public static final int INTAKE_MOTOR_ONE = 0;
+        public static final int INTAKE_MOTOR_TWO = 0;
```
Comment (pordonj) [inline src/main/java/frc/robot/Constants.java:63]: I am initially assigning ID's 9 and 10 to undertaker. Please change these

1-8 swerve
9-10 undertaker
11-12 arm
13-16 shooter

File: src/main/java/frc/robot/Undertaker.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    
+    private CANSparkMax intakeMotor = new CANSparkMax( Constants.Undertaker.INTAKE_MOTOR_ONE, MotorType.kBrushless);
+    private CANSparkMax intakeMotor2 = new CANSparkMax( Constants.Undertaker.INTAKE_MOTOR_TWO, MotorType.kBrushless);
+    
+    public void undertakerStop(){
+        intakeMotor.set(0);
+        intakeMotor2.set(0);
+    }
+    
+      public void noteRelease(double speed){
```
Comment (pordonj) [inline src/main/java/frc/robot/Undertaker.java:18]: let's rename this `run` and combine the 3 methods. Different commands are what should tell it to run at different speeds to intake, eject, and stop.

File: src/main/java/frc/robot/Undertaker.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package frc.robot;
+
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.CANSparkLowLevel.MotorType;
+
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class Undertaker extends SubsystemBase{
+    
+    private CANSparkMax intakeMotor = new CANSparkMax( Constants.Undertaker.INTAKE_MOTOR_ONE, MotorType.kBrushless);
```
Comment (pordonj) [inline src/main/java/frc/robot/Undertaker.java:10]: lets rename these `intakeMotorOne` and `Two`

**Review by pordonj — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Undertaker.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        intakeMotorOne.set(speed);
+        intakeMotorTwo.set(speed);
+
+        //backwards
+        intakeMotorOne.set(speed * -1);
+        intakeMotorTwo.set(speed * -1);
+    }
+    
+/** 
+      public void noteRelease(double speed){
```
Comment (pordonj) [inline src/main/java/frc/robot/Undertaker.java:33]: As a general rule, we don't want to merge commented out code. It can be safely deleted because if we need to see what it was we can look at the state of the repository at past commits.

TLDR: Delete this commented-out code

File: src/main/java/frc/robot/Undertaker.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        intakeMotorOne.set(0);
+        intakeMotorTwo.set(0);
+
+        //forwards
+        intakeMotorOne.set(speed);
+        intakeMotorTwo.set(speed);
+
+        //backwards
+        intakeMotorOne.set(speed * -1);
+        intakeMotorTwo.set(speed * -1);
```
Comment (pordonj) [inline src/main/java/frc/robot/Undertaker.java:29]: You were on the right track here but think about how you would tell the speed to go negative, positive, or to 0. (It's not a conditional, and the answer might be more simple than you think)

**Review by pordonj — CHANGES_REQUESTED** [review]

**Comment by pordonj** [comment]
current status: code needs to be pushed from whatever laptop @realApostrophes was working on on Saturday 1/27

File: src/main/java/frc/robot/Undertaker.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    private CANSparkMax intakeMotorTwo = new CANSparkMax( Constants.Undertaker.INTAKE_MOTOR_TWO, MotorType.kBrushless);
+
+    public Undertaker(){
+        if (Robot.isSimulation()){
+            REVPhysicsSim.getInstance().addSparkMax(intakeMotorOne, DCMotor.getNEO(1));
+            REVPhysicsSim.getInstance().addSparkMax(intakeMotorTwo, DCMotor.getNEO(1));
+        }
+    }
+
+    public void undertakerControl(double speed){
```
Comment (pordonj) [inline src/main/java/frc/robot/Undertaker.java:22]: lets rename this `run` or `setVoltage` - `undertakerControl` is kind of redundant. you should be able to rightclick rename to change it everywhere it is used.

@python-42 @jagernet-ops opinions on name?

File: src/main/java/frc/robot/Constants.java [hunk 7]
```diff
@@ -57,4 +57,13 @@ public class Controller {
         public static final int LEFT_JOYSTICK_BUTTON = 9;
         public static final int RIGHT_JOYSTICK_BUTTON = 10;
     }
+
+    public class Undertaker {
+        public static final int INTAKE_MOTOR_ONE = 13;
+        public static final int INTAKE_MOTOR_TWO = 14;
+
+        public static final double INTAKE_SPEED = 6.0;
```
Comment (pordonj) [inline src/main/java/frc/robot/Constants.java:65]: lets rename these `INTAKE_VOLTAGE` and `EJECT_VOLTAGE` to better describe what they are

File: src/main/java/frc/robot/command/undertaker/EjectNote.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.Constants;
+import frc.robot.Undertaker;
+
+public class EjectNote extends Command {
+    
+     private Undertaker undertaker;
+
+
+  /** Creates a new IntakeNote. */
```
Comment (pordonj) [inline src/main/java/frc/robot/command/undertaker/EjectNote.java:12]: CPFTM - should be `EjectNote`

Comment (pordonj) [reply src/main/java/frc/robot/command/undertaker/EjectNote.java:12]: You could also remove this comment and the matching on in `IntakeNote`, i think they are unnecessary

**Review by pordonj — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Constants.java [hunk 9]
```diff
@@ -57,4 +57,13 @@ public class Controller {
         public static final int LEFT_JOYSTICK_BUTTON = 9;
         public static final int RIGHT_JOYSTICK_BUTTON = 10;
     }
+
+    public class Undertaker {
+        public static final int INTAKE_MOTOR_ONE = 13;
+        public static final int INTAKE_MOTOR_TWO = 14;
+
+        public static final double INTAKE_VOLTAGE = 6.0;
```
Comment (pordonj) [inline src/main/java/frc/robot/Constants.java:65]: after discussion with @python-42 

let's change this back to `INTAKE_SPEED` and be -1 to 1 (so this would be 0.5). Then multiply by 12 in the `Undertaker` subsystem method

Same with eject

File: src/main/java/frc/robot/Undertaker.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    public Undertaker(){
+        if (Robot.isSimulation()){
+            REVPhysicsSim.getInstance().addSparkMax(intakeMotorOne, DCMotor.getNEO(1));
+            REVPhysicsSim.getInstance().addSparkMax(intakeMotorTwo, DCMotor.getNEO(1));
+        }
+    }
+
+    public void run(double speed){
+        intakeMotorOne.setVoltage(speed);
+        intakeMotorTwo.setVoltage(speed);
```
Comment (pordonj) [inline src/main/java/frc/robot/Undertaker.java:24]: Multiply by 12 here per above comment

**Review by pordonj — CHANGES_REQUESTED** [review]
